crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2366 @@
+        return 1;
+    return 0;
+}
+
+/*
+ * @brief generate a key pair keeping private inside API
+ *
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @param mode is the mode (currently unused)
+ * @param suite is the ciphersuite
+ * @param publen is the size of the public key buffer (exact length on output)
+ * @param pub is the public value
+ * @param priv is the private key pointer
+ * @return 1 for good (OpenSSL style), not 1 for error
+ */
+static int hpke_kg_evp(OSSL_LIB_CTX *libctx, const char *propq,
+                       unsigned int mode, OSSL_HPKE_SUITE suite,
+                       size_t ikmlen, const unsigned char *ikm,
+                       size_t *publen, unsigned char *pub,
+                       EVP_PKEY **priv)
+{
+    int erv = 1; /* Our error return value - 1 is success */
+    EVP_PKEY_CTX *pctx = NULL;
+    EVP_PKEY *skR = NULL;
+    unsigned char *lpub = NULL;
+    size_t lpublen = 0;
+    uint16_t kem_ind = 0;
+    OSSL_PARAM params[3], *p = params;
+
+    if (hpke_suite_check(suite) != 1)
+        return 0;
+    if (pub == NULL || publen == NULL || *publen == 0 || priv == NULL)
+        return 0;
+    if (ikmlen > 0 && ikm == NULL)
+        return 0;
+    if (ikmlen == 0 && ikm != NULL)
+        return 0;
+    kem_ind = kem_iana2index(suite.kem_id);
+    if (kem_ind == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (hpke_kem_id_nist_curve(suite.kem_id) == 1) {
+        *p++ = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME,
+                                                (char *)hpke_kem_tab[kem_ind]
+                                                .groupname, 0);
+        pctx = EVP_PKEY_CTX_new_from_name(libctx, "EC", propq);
+    } else {
+        pctx = EVP_PKEY_CTX_new_from_name(libctx, hpke_kem_tab[kem_ind].keytype,
+                                          propq);
+    }
+    if (pctx == NULL
+        || EVP_PKEY_keygen_init(pctx) <= 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (ikm != NULL) {
+        *p++ = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_DHKEM_IKM,
+                                                 (char *)ikm, ikmlen);
+    }
+    *p = OSSL_PARAM_construct_end();
+    if (EVP_PKEY_CTX_set_params(pctx, params) <= 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (EVP_PKEY_generate(pctx, &skR) <=0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    EVP_PKEY_CTX_free(pctx);
+    pctx = NULL;
+    lpublen = EVP_PKEY_get1_encoded_public_key(skR, &lpub);
+    if (lpub == NULL || lpublen == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (lpublen > *publen) {
+        erv = 0;
```

Comment:
Have a look at the internals of EVP_PKEY_get1_encoded_public_key().. You can probably just call EVP_PKEY_get_octet_string_param() without doing a malloc()/memcpy()..

Reply:
That works for cfrg curves but appears to return compressed points for NIST curves so probably not worthwhile.

Reply:
You know the size of the buffer you need I think..

Reply:
EVP_PKEY_get1_encoded_public_key() is really just a helper function that uses EVP_PKEY_get_octet_string_param

Reply:
well, maybe there's a  way to get the uncompressed public keys, but it's likely as much code as what's there already isn't it?

Reply:
````
EVP_PKEY_get_octet_string_param(skR,
                                                            OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY,
                                                             pub,  *publen, publen)
````

Reply:
Single line without a memcpy :)

Reply:
Fair 'nuff magic param name wins:-)